keps/sig-api-machinery/0032-20181026-flexible-inflight-request-limit.md

```diff
@@ -0,0 +1,143 @@
+
+
+### Algorithm to accept/reject readonly request <a name="algorithm-to-accept-reject-readonly-request"></a>
+
+0. Pick group with max `readonly_request` across user's group as `request_group`, pick group with max `readonly_limit` across user's group as `limit_group`.
+1. If $total `weight` of inflight readonly requests of `requestGroup` > `limit_group.readonly_limit`, REJECT inbound request
+2. Else, if total `weight` of inflight readonly requests of `requestGroup` < `request_group.readonly_request`, ACCEPT inbound request
+3. Else, if (sum of `readonly_request` of all groups + total `weight` of all readonly requests, exceedes appropriate `request_group.readonly_request`) < `readonly_capacity`, ACCEPT inbound request
+4. else, REJECT  inbound request
+
+For mutating requests algorithm is the same.
+
+### How to store limits config <a name="how-to-store-limits-config"></a>
+
+Implement new API resource to store group limits and requests:
+```
+apiVersion: extensions/v1beta1
+kind: InflightLimit
+metadata:
+  name: %USER_GROUP_NAME%
+spec:
+  readonly:
+    limit: %READONLY_TOKENS_LIMIT%
+    request: %READONLY_TOKENS_REQUEST%
+  mutating:
+    limit: %MUTATING_TOKENS_LIMIT%
+    request: %MUTATING_TOKENS_REQUEST%
+status:
+  readonlyInflightTokens: %CURRENT_READONLY_TOKENS_INFLIGHT%
+  mutatingInflightTokens: %CURRENT_MUTATING_TOKENS_INFLIGHT%
+```
+
+Where:
+
+`%USER_GROUP_NAME%` - The name of appropriate user group
+
+`%READONLY_TOKENS_LIMIT%` - `readonly_limit`
+
+`%READONLY_TOKENS_REQUEST%` - `readonly_request`
+
+`%MUTATING_TOKENS_LIMIT%` - `mutating_limit`
+
+`%MUTATING_TOKENS_REQUEST%` - `mutating_request`
+
+`%CURRENT_READONLY_TOKENS_INFLIGHT%` - Total `weight` of inflight readonly requests from the user group
+
+`%CURRENT_MUTATING_TOKENS_INFLIGHT%` - Total `weight` of inflight mutating requests from the user group
+
+### How to determinate request weight<a name="how-to-determinate-request-weight"></a>
+
+Implement request `weight` calculation as a function F(verb_weight, related_resources_count), where verb_weight is a constant for particular verb, related_resources_count -- count of pods, events or services -- according to requested resource type and namespace.
+
+Resources counts could be stored in the map and periodically updated in background like in etcd_object_counts metrics: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd/metrics/metrics.go#L62
+
+
+## Discussionable questions <a name="discussionable-questions"></a>
+
+### How to treat namespace-scoped lists? <a name="how-to-treat-namespace-scoped-lists"></a>
+
+There is an open question how to treat namespace-scoped lists (e.g. list all pods from namespace X) - those are clearly different than singular gets and listing all objects of a given type
+
+(@wojtek-t)
+
+**My opinion:** Due to object key schema in etcd (`/registry/%OBJECT_TYPE%/%NAMESPACE%/%OBJECT_NAME%`) there is no difference between
```

Comment:
Thank you for your answer.

> So conceptually they are the same because they depend on the number of corresponding objects. But their cost is not the same.

I meant exactly the same. So, is it OK, if i remove this question?